[AArch64][SME] Improve codegen for aarch64.sme.cnts* when not in streaming mode

kmclaughlin-arm opened the pull request:
Builtins for reading the streaming vector length are canonicalised to
use the aarch64.sme.cntsd intrinisic and a multiply, i.e.
  - cntsb -> cntsd * 8
  - cntsh -> cntsd * 4
  - cntsw -> cntsd * 2

This patch also removes the LLVM intrinsics for cnts[b,h,w], and adds
patterns to improve codegen when cntsd is multiplied by a constant.

kmclaughlin-arm committed:
[AArch64][SME] Improve codegen for aarch64.sme.cnts* when not in streaming mode

kmclaughlin-arm committed:
RDSVL tests

kmclaughlin-arm committed:
- Replace cnts[b|h|w] builtins with cntsd intrinsic in Clang
- Remove cnts[b|h|w] intrinsics in LLVM
- Add patterns for cntsd

paulwalker-arm commented on clang/include/clang/Basic/arm_sme.td:
```diff
@@ -156,16 +156,15 @@ let SMETargetGuard = "sme2p1" in {
-defm SVCNTSH : ZACount<"cntsh">;
-defm SVCNTSW : ZACount<"cntsw">;
-defm SVCNTSD : ZACount<"cntsd">;
+defm SVCNTSB : ZACount<"", "cntsb">;
+defm SVCNTSH : ZACount<"", "cntsh">;
+defm SVCNTSW : ZACount<"", "cntsw">;
+defm SVCNTSD : ZACount<"aarch64_sme_cntsd", "cntsd">;
```
I've changed this code locally several times but it has never ended up in a PR so if you don't mind I'll ask you to do it.  Basically, I hate the `ZACount` class because it offers no value and is inconsistent with how all the other builtins are declared.  Do you mind removing it and just declaring `def SVCNTSB` and friends using `SInst` directly?

kmclaughlin-arm committed:
- Remove cnts[b,h,w] intrinsics from MLIR and fix tests
- Remove ZACount class from arm_sme.td

MacDue commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -845,15 +847,25 @@ struct StreamingVLOpConversion
-        return arm_sme::aarch64_sme_cntsb::create(rewriter, loc, i64Type);
-      case arm_sme::TypeSize::Half:
-        return arm_sme::aarch64_sme_cntsh::create(rewriter, loc, i64Type);
-      case arm_sme::TypeSize::Word:
-        return arm_sme::aarch64_sme_cntsw::create(rewriter, loc, i64Type);
+      case arm_sme::TypeSize::Byte: {
+        auto mul = arith::ConstantIndexOp::create(rewriter, loc, 8);
+        auto mul64 = arith::IndexCastOp::create(rewriter, loc, i64Type, mul);
+        return arith::MulIOp::create(rewriter, loc, cntsd, mul64);
+      }
+      case arm_sme::TypeSize::Half: {
+        auto mul = arith::ConstantIndexOp::create(rewriter, loc, 4);
+        auto mul64 = arith::IndexCastOp::create(rewriter, loc, i64Type, mul);
+        return arith::MulIOp::create(rewriter, loc, cntsd, mul64);
+      }
+      case arm_sme::TypeSize::Word: {
+        auto mul = arith::ConstantIndexOp::create(rewriter, loc, 2);
+        auto mul64 = arith::IndexCastOp::create(rewriter, loc, i64Type, mul);
+        return arith::MulIOp::create(rewriter, loc, cntsd, mul64);
+      }
       case arm_sme::TypeSize::Double:
-        return arm_sme::aarch64_sme_cntsd::create(rewriter, loc, i64Type);
+        return cntsd;
       }
       llvm_unreachable("unknown type size in StreamingVLOpConversion");
     }();
```
nit: I think it's best to get rid of this lambda now and switch and just handle this inline like:
 ```c++    
     auto scale = arith::ConstantIndexOp::create(rewriter, loc, getSizeInBytes(streamingVlOp.getTypeSize()));
     auto mul = arith::IndexCastOp::create(rewriter, loc, i64Type, scale);
     rewriter.replaceOpWithNewOp<arith::IndexCastOp>(streamingVlOp, rewriter.getIndexType(), mul);
```
`getSizeInBytes` would be a static helper that switches on `arm_sme::TypeSize` to return the size in bytes.

MacDue commented on clang/lib/CodeGen/TargetBuiltins/ARM.cpp:
```diff
@@ -4918,6 +4919,31 @@ Value *CodeGenFunction::EmitAArch64SMEBuiltinExpr(unsigned BuiltinID,
   // Handle builtins which require their multi-vector operands to be swapped
   swapCommutativeSMEOperands(BuiltinID, Ops);
 
+  auto isCntsBuiltin = [&](int64_t &Mul) {
+    switch (BuiltinID) {
```
nit: This could just return the `mul`, and the use below could be `if (auto mul = isCntsBuiltin()))`

MacDue commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2825,13 +2825,7 @@ AArch64TTIImpl::instCombineIntrinsic(InstCombiner &IC,
-    return instCombineSMECntsElts(IC, II, 8, ST);
-  case Intrinsic::aarch64_sme_cntsb:
-    return instCombineSMECntsElts(IC, II, 16, ST);
+    return instCombineSMECntsElts(IC, II, ST);
```
Rename this to `instCombineSMECntsd` now that it only handles a single instruction?
```suggestion
    return instCombineSMECntsd(IC, II, ST);
```

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -937,6 +940,23 @@ bool AArch64DAGToDAGISel::SelectRDVLImm(SDValue N, SDValue &Imm) {
+bool AArch64DAGToDAGISel::SelectRDSVLShiftImm(SDValue N, SDValue &Imm) {
+  if (!isa<ConstantSDNode>(N))
+    return false;
+
```
Could you add a comment like:
```
cntsd = (rdvl, #1) >> 3

So, if we have (cntsd << #ShlImm), where ShlImm >= 3, attempt to fold into:

rdvl, #(ShlImm - 3) 
```

MacDue reviewed (commented):
A few little comments:

kmclaughlin-arm committed:
- Remove lambda from StreamingVLOpConversion
- Add getSizeInBytes helper

kmclaughlin-arm committed:
- Fix 'default label in switch' build failure

MacDue reviewed (approved):
Generally LGTM (mainly looking at the MLIR side), just a few final comments:

paulwalker-arm commented on clang/test/CodeGen/AArch64/sme-intrinsics/acle_sme_cnt.c:
```diff
@@ -6,49 +6,55 @@
 #include <arm_sme.h>
 
-// CHECK-C-LABEL: define dso_local i64 @test_svcntsb(
+// CHECK-C-LABEL: define dso_local range(i64 0, -9223372036854775808) i64 @test_svcntsb(
```
Why wrong? `range(i64 0, -9223372036854775808)` says the result is known positive? which is true.

MacDue replied:
I think you're right... I just read `range(i64 0, -9223372036854775808)` as `value >= 0 && < -9223372036854775808` which is wrong, but it seems the range comparisons are unsigned despite how it's printed.

kmclaughlin-arm committed:
- Fix comments in AArch64ISelDAGToDAG & ArmSMEToLLVM

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -937,6 +940,26 @@ bool AArch64DAGToDAGISel::SelectRDVLImm(SDValue N, SDValue &Imm) {
 
+// Given cntsd = (rdsvl, #1) >> 3, attempt to return a suitable multiplier
+// for RDSVL to calculate the streaming vector length in bytes * N. i.e.
+//   rdsvl, #(ShlImm - 3)
```
Realized I messed up that suggestion, `rdsvl, #(N - 3)` should be `rdsvl, #(1 << (N - 3))`

kmclaughlin-arm committed:
- Fix comment in SelectRDSVLShiftImm

kmclaughlin-arm committed:
- Add lowering of cntsd back into LowerINTRINSIC_WO_CHAIN
- Remove patterns for cntsd and add for AArch64rdsvl
- Add nsw/nuw flags to mul of cntsd in EmitSMELd1St1

kmclaughlin-arm committed:
- Remove sme_cnts*_mul_imm patterns

kmclaughlin-arm committed:
- Remove setNoSignedWrap/setNoUnsignedWrap from cntsd lowering

kmclaughlin-arm committed:
- Pass SDNodeFlags::Exact to getNode and remove Flags
